### PR TITLE
Change 'force_recovery' option behavior and makes tarantool to ignore some errors during snapshot recovery just like during xlog recovery

### PR DESCRIPTION
**Pull request by mechanik20051988**
Change 'force_recovery' option behavior and makes tarantool to ignore
some errors during snapshot recovery just like during xlog recovery.

### CODE REVIEW

**Comment by Gerold103** [comment]
What is wrong with using the existing option force_recovery?

**Comment by mechanik20051988** [comment]
> What is wrong with using the existing option force_recovery?

force_recovery option allow to ignore problems during xlog recovery (see #5422), not in snapshot recovery. I add option which allow to ignore problems during snapshot recovery (except damage system spaces or snapshot meta info), such as:
1. garbage at the end of snapshot file
2. truncated snapshot file
3. some garbage at the middle of snapshot file

**Comment by Gerold103** [comment]
I know that it was used for xlogs only. What is the problem to reuse it for snapshots? 'Recovery' includes snapshot loading.

**Comment by kostja** [comment]
* mechanik20051988 <notifications@github.com> [20/10/27 14:19]:
> There was an option &#39;force_recovery&#39; that makes tarantool
> to ignore some problems during xlog recovery. This patch
> add new option &#39;ignore_snap_errors&#39; that makes tarantool
> to ignore errors in snapshot recovery except errors in
> snapshot meta information. This option must used together
> with &#39;force_recovery&#39; option or it will be ignored
> You can view, comment on, or merge this pull request online at:
> 
>   https://github.com/tarantool/tarantool/pull/5443

there is wal built-in module for handling arbitrarily broken 
snapshots and wals.

instead of adding a yet another product option you could have a
standalone tarantool program which would load all kinds of broken
objects into an existing instance.

-- 
Konstantin Osipov, Moscow, Russia

**Comment by kostja** [comment]
* mechanik20051988 <notifications@github.com> [20/10/27 16:10]:
> > What is wrong with using the existing option force_recovery?
> 
> force_recovery option allow to ignore problems during xlog recovery (see #5422), not in snapshot recovery. I add option which allow to ignore problems during snapshot recovery (except damage system spaces or snapshot meta info), such as:
> 1. garbage at the end of snapshot file
> 2. truncated snapshot file
> 3. some garbage at the middle of snapshot file

It actually does fix issues in snapshot recovery as well, just not
all of them.


-- 
Konstantin Osipov, Moscow, Russia

**Comment by kostja** [comment]
We used to have panic_on_wal_error and panic_on_snap_error.
The potions were deprecated since they are impossible to make work
right.

This option is also impossible to make work right - one needs an
external tool.

This is why the options were deprecated in the first place.

File: src/box/memtx_engine.c [hunk 1]
```diff
@@ -1064,6 +1072,9 @@ memtx_engine_new(const char *snap_dirname, bool force_recovery,
 	xdir_create(&memtx->snap_dir, snap_dirname, SNAP, &INSTANCE_UUID,
 		    &xlog_opts_default);
 	memtx->snap_dir.force_recovery = force_recovery;
+	/* This options is useful, only if force_recovery is true */
+	memtx->force_broken_snapshot_recovery =
+		force_recovery ? force_broken_snapshot_recovery : false;
```
Comment (Totktonada) [inline src/box/memtx_engine.c:1077]: So the option will be silently ignored if force_recovery is not set? Will not better to give an explicit error for incorrect parameters?

**Comment by Gerold103** [comment]
I see the branch is updated, but the questions are ignored. Again - why can't force_recovery be just used for all snapshot errors as well? Loading a snapshot is also a part of recovery.

**Comment by mechanik20051988** [comment]
> I see the branch is updated, but the questions are ignored. Again - why can't force_recovery be just used for all snapshot errors as well? Loading a snapshot is also a part of recovery.

Sorry! It was task to me to add new option instead of changing the behavior for the option 'force_recovery'

File: src/box/memtx_engine.c (hunk 1 above)
Comment (mechanik20051988) [reply src/box/memtx_engine.c:1077]: Ok, i add message about error in options

**Comment by Mons** [comment]
1. This behaviour should respect existing option: `force_recovery`
2. Instance should start only in 2 cases:
    * snapshot is someway truncated, but after necessary system spaces
    * single tuple within snapshot has broken checksum and may be skipped without consequences. 
In any other cases instance should not start

Creation of recovery utility is a subject of another issue.

**Comment by kostja** [comment]
@mons this is already how force_recovery works.

**Comment by Mons** [comment]
> @Mons this is already how force_recovery works.

@kostja this is how it should work, but does not. `force_recovery` now affects only `.xlog`'s, not `.snap`'s
So, this is an issue to fix it

**Comment by kostja** [comment]
force_recovery option is passed to the engine, it's up to the engine what to do with it. you're right it's dangerous to recover with a corrupt data dictionary.

File: src/box/memtx_engine.c [hunk 2]
```diff
@@ -78,6 +78,117 @@ enum {
...
+	/* Space _func recover complete index */
+	FUNC_RECOVER_COMPLETE_ID,
+	/* Space _user recover complete index */
+	USER_RECOVER_COMPLETE_ID,
+	/* Space _priv recover complete index */
+	PRIV_RECOVER_COMPLETE_ID,
+	/* Space _cluster recover complete index */
+	CLUSTER_RECOVER_COMPLETE_ID,
+	/* End of enum */
+	RECOVER_COMPETE_MAX_ID,
```
Comment (alyapunov) [inline src/box/memtx_engine.c:99]: I see only one method of check that a space with some ID was fully loaded - if a read a row with space_id > ID. Thus we can ensure that all system spaces was fully loaded if we read a row with space_id = BOX_SYSTEM_ID_MAX + 1. And if we read such a row - that means that we have read ALL system spaces and thus we should not care about each of them individually.

**Review by alyapunov — CHANGES_REQUESTED** [review]
Maybe I just didn't understand the idea but I think the path is wrong. I think we should not track each system space individually and can divide a snapshot into system and non-system parts by the first row with non-system space_id. That would be more simple and frexible.

File: src/box/memtx_engine.c [hunk 3]
```diff
@@ -230,6 +243,8 @@ memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
 	struct request request;
 	if (xrow_decode_dml(row, &request, dml_request_key_map(row->type)) != 0)
 		return -1;
+	if (not_system_space && request.space_id > BOX_SYSTEM_ID_MAX)
```
Comment (alyapunov) [inline src/box/memtx_engine.c:246]: 1. By our code style we should never implicitly convert pointer to bool, we should explicitly compare it with NULL: `not_system_space != NULL`.
2. I don't like that if `request.space_id <= BOX_SYSTEM_ID_MAX` you don't assign the value to false.

File: src/box/memtx_engine.c [hunk 4]
```diff
@@ -216,7 +229,7 @@ memtx_engine_recover_raft(const struct xrow_header *row)
 
 static int
 memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
-				  struct xrow_header *row)
+				  struct xrow_header *row, bool *not_system_space)
```
Comment (alyapunov) [inline src/box/memtx_engine.c:232]: By out code style bool variables should start with `is_`, `was_` etc.

**Review by alyapunov — CHANGES_REQUESTED** [review]

File: src/box/memtx_engine.c (hunk 3 above)
Comment (mechanik20051988) [reply src/box/memtx_engine.c]: Fix

File: src/box/memtx_engine.c (hunk 4 above)
Comment (mechanik20051988) [reply src/box/memtx_engine.c]: Fix

**Review by alyapunov — APPROVED** [review]
LGTM

File: src/box/memtx_engine.c [hunk 5]
```diff
@@ -159,8 +159,16 @@ memtx_engine_recover_snapshot(struct memtx_engine *memtx,
...
+	char filename[PATH_MAX];
+	const char *name = xdir_format_filename(&memtx->snap_dir,
 						    signature, NONE);
+	/*
+	 * we need to save name in local variable because
+	 * xdir_format_filename allocate memory in static buffer
+	 * which will be overwritten later
+	 */
+	strncpy(filename, name, PATH_MAX - 1);
+	filename[PATH_MAX - 1] = '\0';
```
Comment (Korablev77) [inline src/box/memtx_engine.c:171]: As I understand, it is separate bug which is not connected with purpose of this patch. So I suggest you to move this change in a separate patch and also verify other xdir_format_filename() usages. Nit: start comment with a capital letter and end it with a dot.

File: test/box/broken_snapshot.test.lua [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+
+test_run:cmd("create server test with script='box/broken_snapshot.lua'")
+test_run:cmd("start server test")
+test_run:cmd("switch test")
+space = box.schema.space.create('test', { engine = "memtx" })
+space:format({ {name = 'id', type = 'unsigned'}, {name = 'year', type = 'unsigned'} })
+index = space:create_index('primary', { parts = {'id'} })
+
+for key = 1, 10000 do space:insert({key, key + 1000}) end
+box.snapshot()
```
Comment (Korablev77) [inline test/box/broken_snapshot.test.lua:34]: Instead of generating .snap file on each test run, mb it is worth creating it once and put it near test? E.g. see upgrage tests. It would speed up test a little.

File: test/box/broken_snapshot.test.lua [hunk 7]
```diff
@@ -0,0 +1,123 @@
+
```
Comment (Korablev77) [inline test/box/broken_snapshot.test.lua:1]: As a rule we add gh-xxxx- prefix to test name

File: test/box/broken_snapshot.test.lua [hunk 8]
```diff
@@ -0,0 +1,123 @@
...
+test_run:cmd("switch default")
+snapfile = get_snap_file()
+file = io.open(snapfile, "r")
+size = file:seek("end")
+if size > 30000 then size = 30000 end
+
+-- save last snapshot
+os.execute(string.format('cp %s %s.save', snapfile, snapfile))
+-- write garbage at the end of file
+file = io.open(snapfile, "ab")
```
Comment (Korablev77) [inline test/box/broken_snapshot.test.lua:45]: What about more sophisticated snap file contamination? Like change space id to non-existent, insert garbage in the middle of snap, corrupt system space etc

File: test/box/broken_snapshot.test.lua [hunk 9]
```diff
@@ -0,0 +1,123 @@
...
+file:seek("set", size)
+for i = 1, 1000, 1 do file:write(math.random(0,255)) end
+test_run:cmd("switch test")
+test_run:cmd("restart server test with script='box/broken_snapshot.lua'")
+test_run:cmd("setopt delimiter ';'")
+-- check that some data valid
+val = box.space.test:select();
+for i = 1, 1000, 1 do
+    if (val[i] == nil) then
+        os.exit()
```
Comment (Korablev77) [inline test/box/broken_snapshot.test.lua:91]: Why os.exit()? Are all these 1k rows should be recovered? It should be deterministic test, so you can place asserts I guess.

File: src/box/memtx_engine.c [hunk 10]
```diff
@@ -170,12 +178,13 @@ memtx_engine_recover_snapshot(struct memtx_engine *memtx,
 	int rc;
 	struct xrow_header row;
 	uint64_t row_count = 0;
+	int is_not_system_space = -1;
```
Comment (Korablev77) [inline src/box/memtx_engine.c:181]: Why don't use boolean type?

Comment (Korablev77) [reply src/box/memtx_engine.c:181]: Also would be nice to see explanation why we have to stop recovery if one of system space's rows is broken.

File: src/box/memtx_engine.c [hunk 11]
```diff
@@ -230,6 +243,9 @@ memtx_engine_recover_snapshot_row(struct memtx_engine *memtx,
 	struct request request;
 	if (xrow_decode_dml(row, &request, dml_request_key_map(row->type)) != 0)
 		return -1;
+	if (is_not_system_space != NULL)
```
Comment (Korablev77) [inline src/box/memtx_engine.c:246]: I would always pass non-null pointer and assign is_not_system_space. Also I'd rename the latter to is_space_system. Finally, you can use space_is_system() function.

File: src/box/memtx_engine.c (hunk 10 above)
Comment (mechanik20051988) [reply src/box/memtx_engine.c:181]: We can't use boolean type here because there is a case where xlog_cursor_next return not 0 imidiatly. This means that snapshoot system data is strpongly corrupted. in this case we check that is_not_system_space  < 0 later (it's not changed and == -1) and  return from this function with fail.

File: test/box/broken_snapshot.test.lua (hunk 8 above)
Comment (mechanik20051988) [reply test/box/broken_snapshot.test.lua]: about system space Mons write in some comment before that we not need to boot when system space was corrupted, about change space id, I try to add this case in test

File: test/box/broken_snapshot.test.lua (hunk 7 above)
Comment (mechanik20051988) [reply test/box/broken_snapshot.test.lua]: ok

File: src/box/memtx_engine.c [hunk 12]
```diff
@@ -178,12 +178,16 @@ memtx_engine_recover_snapshot(struct memtx_engine *memtx,
 	int rc;
 	struct xrow_header row;
 	uint64_t row_count = 0;
+	int is_space_system = -1;
 	while ((rc = xlog_cursor_next(&cursor, &row,
-				      memtx->force_recovery)) == 0) {
+		/* In case when we read system space, we can't ignore errors */
```
Comment (Korablev77) [inline src/box/memtx_engine.c:183]: Basically, you can factor out first xlog_cursor_next() call since it is the only case when we substitute memtx->force_recovery with false (in all next calls to xlog_cursor_next() is_space_system will be either false or true).
I've reverted this change and all tests seem to pass. Could you add test covering this case?
Also we do not leave comments between arguments and it is not preferable to evaluate non-trivial condition in-place of argument.

File: test/box/gh-5422-broken_snapshot.test.lua [hunk 13]
```diff
@@ -0,0 +1,124 @@
...
+os.execute(string.format('rm %s.save', snapfile))
+-- write big garbage at the start of file
+file = io.open(snapfile, "r+b")
+file:seek("set", size / 2)
+for i = 1, 1000, 1 do file:write(math.random(1,254)) end
+io.close(file)
+test_run:cmd("start server test with crash_expected=True")
+log = string.format("%s/%s.%s", fio.cwd(), "gh-5422-broken_snapshot", "log")
+-- We must not find ER_UNKNOWN_REPLICA in log file, so grep return not 0
+assert(os.execute(string.format("cat %s | grep ER_UNKNOWN_REPLICA:", log)) ~= 0)
```
Comment (Korablev77) [inline test/box/gh-5422-broken_snapshot.test.lua:122]: You can use test_run:grep_log() instead.

File: test/box/gh-5422-broken_snapshot.test.lua [hunk 14]
```diff
@@ -0,0 +1,124 @@
...
+val = box.space.test:select();
+for i = 1, 1000, 1 do
+    assert(val[i] ~= nil)
+end;
+test_run:cmd("setopt delimiter ''");
+test_run:cmd("switch default")
+test_run:cmd('stop server test')
+
+-- restore snapshot
+os.execute(string.format('cp %s.save %s', snapfile, snapfile))
```
Comment (Korablev77) [inline test/box/gh-5422-broken_snapshot.test.lua:112]: I don't see that you delete original snapfile at the end of test: only snapfile.save is deleted. Please make sure that no artifacts after run are remained (I haven't checked, so it is rather question).

Comment (mechanik20051988) [reply test/box/gh-5422-broken_snapshot.test.lua:112]: It's delete by 
test_run:cmd("cleanup server test")
test_run:cmd("delete server test")
ls -la var/001_box/gh-5422-broken_snapshot
итого 8
drwxr-x--- 2 root root 4096 дек 22 17:46 .
drwxr-xr-x 5 root root 4096 дек 22 17:46 ..

File: test/box/gh-5422-broken_snapshot.test.lua (hunk 13 above)
Comment (mechanik20051988) [reply test/box/gh-5422-broken_snapshot.test.lua:122]: As i see grep_log hangs if the server is dead, by the time it is called

**Review by Korablev77 — APPROVED** [review]
LGTM. Pushed to master, changelog is updated.
